### ISSUE TITLE
Client not respecting schema types

### ISSUE THREAD

**Issue by zshannon** [prisma/prisma#25365]
### Bug description

In my prisma schema, I have a type like:

```prisma
model Property {
    key String
    value String 
}
```

If I store a value like `"2024-10-04T01:23:56.900+00:00"` in `value`, I subsequently get errors from the prisma client when I attempt to store other string values in `value` because the client has decided `value` is a DateTime.

### How to reproduce

1. Create a model with a String field.
2. Store an ISO-8601 compatible string in it.
3. Attempt to store a string that is not ISO-8601 compatible in it.

### Expected behavior

The prisma client should respect the types defined in the prisma schema.

### Prisma information

```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS: macOS
- Database: SQLite
- Node.js version: v20.11.0

### Prisma Version

```
prisma                  : 5.20.0
@prisma/client          : 5.20.0
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v20.11.0
Query Engine (Node-API) : libquery-engine 06fc58a368dc7be9fbbbe894adf8d445d208c284 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli 06fc58a368dc7be9fbbbe894adf8d445d208c284 (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.20.0-12.06fc58a368dc7be9fbbbe894adf8d445d208c284
Default Engines Hash    : 06fc58a368dc7be9fbbbe894adf8d445d208c284
Studio                  : 0.502.0
Preview Features        : driverAdapters
```

**Comment by SevInf** [comment]
Hi @zshannon.
Unfortunately, I can not reproduce your issue:

```prisma
datasource db {
  provider = "sqlite"
  url      = "file:dev.db"
}

model Property {
  id    Int    @id @default(autoincrement())
  key   String
  value String
}
```

```ts
  await prisma.property.deleteMany()
  await prisma.property.create({
    data: {
      key: 'foo',
      value: '2024-10-04T01:23:56.900+00:00',
    },
  })

  await prisma.property.create({
    data: {
      key: 'bar',
      value: 'not date',
    },
  })

  const props = await prisma.property.findMany()
  console.log(props)
  ```
  
  this logs:
  
 ```
[
  { id: 3, key: 'foo', value: '2024-10-04T01:23:56.900+00:00' },
  { id: 4, key: 'bar', value: 'not date' }
]
```

could you please provide repository where we could see the error happen?

**Comment by SevInf** [comment]
Hey @zshannon  ,

Did you get a chance to have a look at my previous comment? It'd be helpful for us to move your issue forward.

Without further information, we'll have to close this issue. Thank you 🙏

**Comment by zshannon** [comment]
#25404 is another example of this

**Comment by zshannon** [comment]
And now I can't get close to reproducing this for you because there's a new worse bug which you can find in this 😬  [repo](https://github.com/zshannon/prisma-d1-perf-benchmark)

<img width="1161" alt="Screenshot 2024-10-16 at 12 17 59 PM" src="https://github.com/user-attachments/assets/ab343fe1-cadc-4b95-a301-68cd43653e38">

**Comment by zshannon** [comment]
okay here's the [repro](https://github.com/zshannon/prisma-d1-perf-benchmark)

tldr: bug doesn't appear when creating directly, but it does appear at least when creating nested objects like.

```prisma
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  previewFeatures = ["driverAdapters"]
  provider        = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model Property {
  id       Int     @id @default(autoincrement())
  key      String
  value    String
  parent   Parent? @relation(fields: [parentId], references: [id])
  parentId Int?
}

model Parent {
  id Int @id @default(autoincrement())

  properties Property[]
}
```

```typescript
export async function writeProperties(count: number) {
	await Promise.all(
		Array.from({ length: count })
			.fill(null)
			.map(async () =>
				prisma.property.create({ data: { key: createId(), value: createId() } })
			)
	)
	await Promise.all(
		Array.from({ length: count })
			.fill(null)
			.map(async () =>
				prisma.parent.create({
					data: {
						properties: {
							createMany: { data: [{ key: createId(), value: DateTime.utc().toISO() }] },
						},
					},
					include: { properties: true },
				})
			)
	)
	revalidatePath('/')
}
```

<img width="1028" alt="Screenshot 2024-10-16 at 12 59 52 PM" src="https://github.com/user-attachments/assets/714d4a23-1002-4e75-852d-85f03a43ed88">

**Comment by zshannon** [comment]
an interesting thing of note here is the db write succeeds despite the error 😬 (take note of the `count` value, click the button, see the error, hit refresh, see the count increased)

**Comment by zshannon** [comment]
okay so i don't understand the purpose of these inferred column types, but i have found an incongruity between the regexes for dates that points to a reasonable compromise.

`sqliteDateRegex` is boxed in with `^` and `$` but `isoDateRegex` is not and at the very least it seems like they both should be and this change should ship right away.

[packages/adapter-d1/src/conversion.ts:55](https://github.com/prisma/prisma/blob/6819678dd575677ccd5c6a72ea51f8421f6a82ce/packages/adapter-d1/src/conversion.ts#L55)
```typescript
// See https://stackoverflow.com/a/3143231/1345244
const isoDateRegex = new RegExp(
  /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
)

// SQLITE date format, returned by built-in time functions. See https://www.sqlite.org/lang_datefunc.html
// Essentially, it is "<ISO Date> <ISO Time>"
const sqliteDateRegex = /^\d{4}-[0-1]\d-[0-3]\d [0-2]\d:[0-5]\d:[0-5]\d$/
function isISODate(str: string) {
  return isoDateRegex.test(str) || sqliteDateRegex.test(str)
}

function inferStringType(value: string): ColumnType {
  if (isISODate(value)) {
    return ColumnTypeEnum.DateTime
  }

  return ColumnTypeEnum.Text
}
```

**Comment by DrJume** [comment]
We have a very similar issue...
We have a field named dateIso which is type String.
Setting an ISO Date works, but querying it with gte/lte throws the error "Invalid value for argument lte: premature end of input. Expected ISO-8601 DateTime."

It should just do string comparison 🤷 no DateTime conversion.

**Comment by zshannon** [comment]
i can't wrap my head around why this the d1 adaptor doesn't use the type information available at compile time and instead relies on runtime inference with terrible performance and correctness outcomes.

it's trivial to add a prisma generator and connect type information from the prisma schema to the adaptor. prisma's sqlite engine only addresses a single table at a time so mapping back to the prisma models isn't hard.

the design of this adaptor is just wrong.

**Comment by acidjazz** [comment]
> okay so i don't understand the purpose of these inferred column types, but i have found an incongruity between the regexes for dates that points to a reasonable compromise.
> 
> `sqliteDateRegex` is boxed in with `^` and `$` but `isoDateRegex` is not and at the very least it seems like they both should be and this change should ship right away.
> 
> [packages/adapter-d1/src/conversion.ts:55](https://github.com/prisma/prisma/blob/6819678dd575677ccd5c6a72ea51f8421f6a82ce/packages/adapter-d1/src/conversion.ts#L55)
> 
> ```ts
> // See https://stackoverflow.com/a/3143231/1345244
> const isoDateRegex = new RegExp(
>   /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
> )
> 
> // SQLITE date format, returned by built-in time functions. See https://www.sqlite.org/lang_datefunc.html
> // Essentially, it is "<ISO Date> <ISO Time>"
> const sqliteDateRegex = /^\d{4}-[0-1]\d-[0-3]\d [0-2]\d:[0-5]\d:[0-5]\d$/
> function isISODate(str: string) {
>   return isoDateRegex.test(str) || sqliteDateRegex.test(str)
> }
> 
> function inferStringType(value: string): ColumnType {
>   if (isISODate(value)) {
>     return ColumnTypeEnum.DateTime
>   }
> 
>   return ColumnTypeEnum.Text
> }
> ```
> 
> adopting this change would un-break the problematic behavior in #25404 where we can at least store ISO strings in JSON objects without throwing an error.

isoDateRegex needs to have ^ and $, i cant store JSON payloads with dates in them because of this.

**Comment by acidjazz** [comment]
Current silly workaround:

```ts
export function useHideIso(input: string, reverse: boolean = false): string {
  if (reverse)
  // Replace FIXME with T
    return input.replace(/(\d{4}-\d{2}-\d{2})FIXME(\d{2}:\d{2}:\d{2}[+-]\d{2}:\d{2})/g, '$1T$2')
  else
  // Replace T with FIXME
    return input.replace(/(\d{4}-\d{2}-\d{2})T(\d{2}:\d{2}:\d{2}[+-]\d{2}:\d{2})/g, '$1FIXME$2')
}
```

**Comment by jkomyno** [comment]
Hi, this issue was fixed by https://github.com/prisma/prisma/pull/25483 in Prisma 5.22.0.

**Comment by acidjazz** [comment]
> Hi, this issue was fixed by #25483 in Prisma 5.22.0.

You're referencing a PR that was added in October, you sure you want to close this even though I can duplicate this error in the latest versions?
